### ISSUE TITLE
Error when using encoder cnn and decoder cnn in opennmt pytorch

### ISSUE THREAD

**Issue by vince62s** [OpenNMT/OpenNMT-py#1323]
@flauted 
I am away, can you please check this one ?

http://forum.opennmt.net/t/error-when-using-encoder-cnn-and-decoder-cnn-in-opennmt-pytorch/2479

**Comment by flauted** [comment]
If I remember right the CNNs have different shaped inputs than what the transformer and RNNs expect. I will look into it.

**Comment by flauted** [comment]
Tl;dr I don't think there's a clear fix for this one... It would require prescribing how the attention in the CNN decoder handles working with a recurrent encoder.

The CNN encoder returns (src word embeddings ``(emb, batch, seq)``, CNN stepwise output ``(cnn, batch, seq)``, lengths) whereas the RNN encs return (hidden vec(s), all hiddens ``memory_bank``, lengths) and the transformer enc returns (src word embeddings, stepwise output, lengths).

The CNN decoder uses (1) ``src_memory_bank_c = self.state["src"] = src word embeddings + CNN output``, the enc outputs with a residual connection to the embeddings, and (2) ``src_memory_bank_t = CNN output``. (1) is used as the keys for attention, (2) is used as the values. Meaning, the attention weights are calculated using the dec state and (1), and those attention weights are used for a weighted sum over (2). See equations 1 and 2 in [2].

I guess for a recurrent encoder, (1) would be the source word embeddings plus the "memory bank" and (2) would just be the "memory bank."

But the code isn't really set up to do that. To me, the clearest way to achieve that would be having all the encoders return their hidden vector(s) (or None in the case of the CNN and transformer), the source embeddings, the "memory bank", and the lengths. And the decoders would have to be updated accordingly. Makes sense since non-recurrent approaches are becoming more popular and they often use some sort of residual connection to the source embeddings.

Another way would be having the encoder return hidden (or None), keys, and values. That's one less return value, but it requires that the encoder knows how the decoder uses its outputs and [maybe attention isn't here to stay](https://arxiv.org/pdf/1901.10430.pdf).

There's a (not nearly as popular) "companion" paper [1] to the fully convolutional one [2] describing how to use a convolutional encoder with a recurrent decoder (which is very different - it uses a separate enc CNN for the keys and the values of attention), but I'm not aware of a recurrent decoder with a CNN encoder.

I wonder if you can plug-and-play a recurrent enc and a convolutional dec in Fairseq.

[1]: "A Convolutional Encoder Model for Neural Machine Translation" https://arxiv.org/pdf/1611.02344.pdf
[2]: "Convolutional Sequence to Sequence Learning" https://arxiv.org/pdf/1705.03122.pdf

----

To the forum poster, is having a convolutional decoder essential to your use-case or will a recurrent one work? If recurrent will work, consider sticking with the train command from the tutorial.

**Comment by vince62s** [comment]
I'm confused, isn't the original post just a cnn dec with cnn enc ? (the plain tuto)

**Comment by flauted** [comment]
The forum post is using data_type img. That overrides whatever encoder is specified and uses the image encoder.

**Comment by vince62s** [comment]
okay I was misled by the fact he said he followed the tutorial.

I am closing this for now since to me it is a much wider problem which is to make encoder and decoder interfaces compatible so that we cam mix different type of architectures.
I already opened a "project" for this subject.

**Comment by StephenKyung** [comment]
@vince62s  
can you share the "project" link?
